examples/mdetr/matcher.py

```diff
@@ -0,0 +1,131 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import List, Tuple
+
+import torch
+import torch.nn.functional as F
+from scipy.optimize import linear_sum_assignment
+from torch import nn, Tensor
+from torchvision.ops.boxes import box_convert, generalized_box_iou
+
+
+class HungarianMatcher(nn.Module):
+    """This class computes an assignment between the targets and the predictions of the network.
+
+    For efficiency reasons, the targets don't include the no_object. Because of this,
+    in general, there are more predictions than targets. In this case, we do a 1-to-1
+    matching of the best predictions, while the others are un-matched (and thus treated
+    as non-objects). This implementation is based on the MDETR repo:
+    https://github.com/ashkamath/mdetr/blob/main/models/matcher.py#L13
+
+    Params: cost_class (float): Relative weight of the classification error in the
+                matching cost. Default: 1
+            cost_bbox (float): Relative weight of the L1 error of the bounding box
+                coordinates in the matching cost. Default: 1
+            cost_giou (float): Relative weight of the giou loss of the bounding box in
+                the matching cost. Default: 1
+
+
+    Inputs: pred_logits (Tensor): Classification logits.
+                Size: (batch_size, num_queries, num_classes)
+            pred_boxes (Tensor): Predicted box coordinates.
+                Size: (batch_size, num_queries, 4)
+            target_boxes_per_sample (List[Tensor]): A list of target bounding boxes.
+                Length = batch_size.
+                Each element is a tensor of size (n_boxes_for_sample, 4).
+            positive_map (Tensor): positive_map[i,j] = 1 when box i maps to class j.
+                Size: (total_boxes, num_classes) where total_boxes is the sum of
+                n_boxes_for_sample over every sample in the batch.
+
+    Returns:
+            A list of size batch_size, containing tuples of (index_i, index_j) where:
+                - index_i is the indices of the selected predictions (in order)
```

Comment:
Leaving out the math formatting for the len(index_i) = len(index_j) ... equation because ultimately I will just wind up wrapping almost everything in \text anyways 😅